### PR TITLE
firewall SSH command

### PR DESCRIPTION
**Pull request by GrigoriyMikhalkin**

### CODE REVIEW

**Comment by GrigoriyMikhalkin** [comment]
Tried to simplify the code and implement SSH connection using [Tailscale's Magic DNS](https://tailscale.com/blog/2021-09-private-dns-with-magicdns). It works when `tailscaled` is running as a daemon on a host. But not when containerized.

There's too much work to adapt `Magic DNS` to work with containerized version. For example `resolv.conf` and routing table has to be updated for that. I think it's better to avoid it since it's not only quite complex but also can cause some problems for the client.

Will return to the not so pretty solution of parsing `tailscale status` response to extract the Firewall's IP address.

**Comment by majst01** [comment]
> Tried to simplify the code and implement SSH connection using [Tailscale's Magic DNS](https://tailscale.com/blog/2021-09-private-dns-with-magicdns). It works when `tailscaled` is running as a daemon on a host. But not when containerized.
> 
> There's too much work to adapt `Magic DNS` to work with containerized version. For example `resolv.conf` and routing table has to be updated for that. I think it's better to avoid it since it's not only quite complex but also can cause some problems for the client.
> 
> Will return to the not so pretty solution of parsing `tailscale status` response to extract the Firewall's IP address.

@GrigoriyMikhalkin there is `tailscale status --json` which makes "parsing" easy, but you already know that for sure

File: cmd/firewall.go [hunk 1]
```diff
@@ -62,7 +73,17 @@ func newFirewallCmd(c *config) *cobra.Command {
...
 	}
 
-	return genericcli.NewCmds(cmdsConfig)
+	firewallSSHCmd := &cobra.Command{
+		Use:   "ssh <firewall ID>",
+		Short: "SSH to a firewall",
+		Long:  `SSH to a firewall via VPN.`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return c.firewallSSH(args)
+		},
```
Comment (Gerrit91) [inline cmd/firewall.go:82]: Auto completion needs to be added

File: cmd/firewall.go [hunk 2]
```diff
@@ -173,3 +194,163 @@ func (c *firewallCmd) createRequestFromCLI() (*models.V1FirewallCreateRequest, e
...
+			"/var/lib:/var/lib",
+			"/dev/net/tun:/dev/net/tun",
+		},
+		NetworkMode: container.NetworkMode("host"),
+		Privileged:  true,
+	}
+	containerName := "tailscaled"
+	resp, err := cli.ContainerCreate(ctx, containerConfig, hostConfig, nil, nil, containerName)
+	if err != nil {
+		return c.describePrinter.Print(err)
```
Comment (Gerrit91) [inline cmd/firewall.go:242]: Just return the error, do not use the printer.

File: cmd/firewall.go [hunk 3]
```diff
@@ -173,3 +194,163 @@ func (c *firewallCmd) createRequestFromCLI() (*models.V1FirewallCreateRequest, e
...
+		)
+	}
+
+	// Connect to the firewall via SSH
+	firewallVPNAddr, err := c.getFirewallVPNAddr(ctx, cli, containerName, *machineGetResp.Payload.Allocation.Hostname)
+	if err != nil {
+		return fmt.Errorf("failed to get Firewall VPN address")
+	}
+
+	err = SSHClient("metal", viper.GetString("identity"), firewallVPNAddr, 22)
```
Comment (Gerrit91) [inline cmd/firewall.go:289]: Identity path needs to be expanded with home directory. Should be somewhere in the project already.

File: cmd/firewall.go [hunk 4]
```diff
@@ -173,3 +194,163 @@ func (c *firewallCmd) createRequestFromCLI() (*models.V1FirewallCreateRequest, e
...
+	}
+
+	cli, err := dockerclient.NewClientWithOpts(dockerclient.FromEnv)
+	if err != nil {
+		return fmt.Errorf("failed to initialize Docker client: %w", err)
+	}
+
+	// Deploy tailscaled
+	ctx := context.Background()
+	containerConfig := &container.Config{
```
Comment (Gerrit91) [inline cmd/firewall.go:227]: Image has to be pulled before starting the container.

File: cmd/firewall.go [hunk 5]
```diff
@@ -173,3 +198,189 @@ func (c *firewallCmd) createRequestFromCLI() (*models.V1FirewallCreateRequest, e
...
+		return fmt.Errorf("failed to pull tailscale image: %w", err)
+	}
+
+	containerConfig := &container.Config{
+		Image: tailscaleImage,
+		Cmd:   []string{"tailscaled"},
+	}
+	hostConfig := &container.HostConfig{
+		Binds: []string{
+			"/var/lib:/var/lib",
```
Comment (majst01) [inline cmd/firewall.go:241]: This is Linux only, actually metalctl also supports macos and windows. We should at least spit out a warning that this command is not available on OSes other than linux. 

On linux a check before if /dev/net/tun is present and give a warning that tun.ko must be loaded.

File: cmd/ssh.go [hunk 6]
```diff
@@ -99,3 +107,16 @@ func publicKey(path string) (ssh.AuthMethod, error) {
 	}
 	return ssh.PublicKeys(signer), nil
 }
+
+func expandFilepath(path string) (string, error) {
```
Comment (majst01) [inline cmd/ssh.go:111]: i think we do this already somewhere in metalctl

**Review by majst01 — APPROVED** [review]

File: cmd/firewall.go (hunk 5 above)
Comment (GrigoriyMikhalkin) [reply cmd/firewall.go]: `/var/lib` looks like was an excessive thing. It was used for storing tailscale state, therefore mounting host directory looks unnecessary. Removed that part, at least during testing it didn't affect anything.

File: cmd/firewall.go [hunk 7]
```diff
@@ -173,3 +202,238 @@ func (c *firewallCmd) createRequestFromCLI() (*models.V1FirewallCreateRequest, e
...
+		return fmt.Errorf("failed to pull tailscale image: %w", err)
+	}
+
+	containerConfig := &container.Config{
+		Image: tailscaleImage,
+		Cmd:   []string{"tailscaled"},
+	}
+	hostConfig := &container.HostConfig{
+		Binds: []string{
+			"/dev/net/tun:/dev/net/tun",
```
Comment (majst01) [inline cmd/firewall.go:295]: Maybe we can also get rid of this by using userspace networking for tailscale as described here:
https://tailscale.com/kb/1112/userspace-networking/

Comment (GrigoriyMikhalkin) [reply cmd/firewall.go]: Done

**Comment by GrigoriyMikhalkin** [comment]
@majst01 The last bit -- removed dependency on `tun` device. Now SSH connection is done via SOCKS5 proxy(ran by `tailscaled` in container). I think it was the last `linux`-only bit in the implementation. Can show the demo on mini-lab. 

Also, can we somehow test `firewall ssh` command on Windows/Mac machines?

**Review by majst01 — APPROVED** [review]

**Review by Gerrit91 — APPROVED** [review]

**Pull request merged**